src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +199,244 @@ bool intersects(Ray const &ray, Box const &box)
+  // will scale with the problem size.
+  float const mag_oA = std::sqrt(dotProduct(oA, oA));
+  float const mag_oB = std::sqrt(dotProduct(oB, oB));
+  float const mag_oC = std::sqrt(dotProduct(oC, oC));
+
+  auto mag_bar = 3.0 / (mag_oA + mag_oB + mag_oC);
+
+  Point A;
+  Point B;
+  Point C;
+
+  // perform shear and scale of vertices
+  // normalized by mag_bar
+  A[0] = (oA[kx] - s[0] * oA[kz]) * mag_bar;
+  A[1] = (oA[ky] - s[1] * oA[kz]) * mag_bar;
+  B[0] = (oB[kx] - s[0] * oB[kz]) * mag_bar;
+  B[1] = (oB[ky] - s[1] * oB[kz]) * mag_bar;
+  C[0] = (oC[kx] - s[0] * oC[kz]) * mag_bar;
+  C[1] = (oC[ky] - s[1] * oC[kz]) * mag_bar;
+
+  // calculate scaled barycentric coordinates
+  float u = C[0] * B[1] - C[1] * B[0];
+  float v = A[0] * C[1] - A[1] * C[0];
+  float w = B[0] * A[1] - B[1] * A[0];
+
+  // fallback to double precision
+  if (u == 0 || v == 0 || w == 0)
+  {
+    u = (double)C[0] * B[1] - (double)C[1] * B[0];
+    v = (double)A[0] * C[1] - (double)A[1] * C[0];
+    w = (double)B[0] * A[1] - (double)B[1] * A[0];
+  }
+
+  constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+  tmin = inf;
+  tmax = -inf;
+
+  // 'Back-face culling' is not supported.
+  // Back-facing culling is to check whether
+  // a surface is 'visible' to a ray, which requires
+  // consistent definition of the facing of triangles.
+  // Once the facing of triangle is defined,
+  // only one of the conditions is needed,
+  // either (u < 0 || v < 0 || w < 0) or
+  // (u > 0 || v > 0 || w > 0), for Back-facing culling.
+  float const epsilon = 0.0000001f;
+  if ((u < -epsilon || v < -epsilon || w < -epsilon) &&
+      (u > epsilon || v > epsilon || w > epsilon))
+    return false;
+
+  // calculate determinant
+  float det = u + v + w;
+
+  A[2] = s[2] * oA[kz];
+  B[2] = s[2] * oB[kz];
+  C[2] = s[2] * oC[kz];
+
+  if (det < -epsilon || det > epsilon)
+  {
+    float t = (u * A[2] + v * B[2] + w * C[2]) / det;
+    tmax = t;
+    tmin = t;
+    return true;
+  }
+
+  // The ray is co-planar to the triangle.
+  // Check the intersection with each edge
+  // the rotate2D function is to make sure the ray-edge
+  // intersection check is at the plane where ray and edges
+  // are at.
+  auto A_star = rotate2D(A);
+  auto B_star = rotate2D(B);
+  auto C_star = rotate2D(C);
+
+  float t_ab = inf;
+  bool ab_intersect = rayEdgeIntersect(A_star, B_star, t_ab);
+  if (ab_intersect)
+  {
+    tmin = t_ab;
+    tmax = t_ab;
+  }
+  float t_bc = inf;
+  bool bc_intersect = rayEdgeIntersect(B_star, C_star, t_bc);
+  if (bc_intersect)
+  {
+    tmin = KokkosExt::min(tmin, t_bc);
+    tmax = KokkosExt::max(tmax, t_bc);
+  }
+  float t_ca = inf;
+  bool ca_intersect = rayEdgeIntersect(C_star, A_star, t_ca);
+  if (ca_intersect)
+  {
+    tmin = KokkosExt::min(tmin, t_ca);
+    tmax = KokkosExt::max(tmax, t_ca);
+  }
+
+  if (ab_intersect || bc_intersect || ca_intersect)
+  {
+    // When (1) the origin of the ray is within the triangle
```

Comment:
Shouldn't it be:
```suggestion
    // When (1) the origin of the ray is on one of the edges of the triangle
```

Reply:
We need to define the cases which the ray origin is inside of the triangle, as well.